libs/storage/IPFS-CLUSTER-DESIGN.md

```diff
@@ -0,0 +1,234 @@
+- Settings with value `0` or empty string use cluster defaults
+
+**Operational Features:**
+- `requestTimeoutMs`: Overrides HTTP timeout for IPFS operations specifically
+- `retryAttempts`: Implements exponential backoff retry logic for failed requests
+- `enableHealthChecks`: Provides cluster health monitoring and status reporting
+
+## Key Implementation Details
+
+**Retry Logic with Exponential Backoff:**
+```typescript
+private async makeRequestWithRetry<T>(
+  requestFn: () => Promise<T>,
+  operation: string,
+): Promise<T> {
+  for (let attempt = 1; attempt <= this.config.retryAttempts; attempt++) {
+    try {
+      return await requestFn();
+    } catch (error: any) {
+      if (attempt === this.config.retryAttempts) throw error;
+      
+      const backoffMs = Math.min(1000 * Math.pow(2, attempt - 1), 10000);
+      await new Promise(resolve => setTimeout(resolve, backoffMs));
+    }
+  }
+}
+```
+
+**Health Check Implementation:**
+```typescript
+async performHealthCheck(): Promise<{ status: string; details: any }> {
+  try {
+    const version = await this.getVersion();
+    return { status: 'healthy', details: version };
+  } catch (error: any) {
+    return { status: 'unhealthy', details: { error: error.message } };
+  }
+}
+```
+
+**Duration to Timestamp Conversion:**
+```typescript
+private parseDurationToTimestamp(duration: string): string {
+  const match = duration.match(/^(\d+)([hdm])$/);
+  if (!match) throw new Error('Invalid duration format');
+  
+  const value = parseInt(match[1], 10);
+  const unit = match[2];
+  const now = new Date();
+  
+  let milliseconds = 0;
+  switch (unit) {
+    case 'h': milliseconds = value * 60 * 60 * 1000; break;
+    case 'd': milliseconds = value * 24 * 60 * 60 * 1000; break;
+    case 'm': milliseconds = value * 60 * 1000; break;
+  }
+  
+  return new Date(now.getTime() + milliseconds).toISOString();
+}
+```
+
+**Pin Status Verification:**
+```typescript
+private isPinnedInCluster(clusterData: any): boolean {
+  if (!clusterData?.peer_map) return false;
+  
+  for (const peerId in clusterData.peer_map) {
+    const peerInfo = clusterData.peer_map[peerId];
+    if (peerInfo?.status === 'pinned' || peerInfo?.status === 'pinning') {
+      return true;
+    }
+  }
+  return false;
+}
+```
+
+**Error Handling:**
+```typescript
+try {
+  const pinStatus = await this.checkPinStatus(cid);
+  return this.isPinnedInCluster(pinStatus.data);
+} catch (err: any) {
+  if (err.message?.includes('not found') || err.message?.includes('404')) {
+    return false; // CID not found = not pinned
+  }
+  throw err; // Re-throw other errors
+}
+```
+
+## Migration Strategy
+
+Existing deployments can migrate to cluster mode by:
+1. **Environment Update**: Set `IPFS_MODE=cluster` and cluster endpoint
+2. **Service Restart**: No code changes required - configuration injected at runtime
+3. **Gradual Migration**: Services can run in mixed mode during transition
```

Comment:
In the interest of accounting for stupidity, we might want to make it explicit that the IPFS URL that you're pointing to must be a cluster 😛 (ie, you can't just activate cluster mode in Gateway and reap the benefits)
```
Step 1: set IPFS_MODE=cluster
Step 2: ?
Step 3: PROFIT!
```